plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +119,36 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			// If the error is an overflow, we probably have an upstream misbehaving in some way.
+			// (e.g. sending >512 byte UDP responses without an eDNS0 OPT RR).
+			// Instead of returning an error, return an empty response with TC bit set. This will make the
+			// client retry over TCP (if that's supported) or at least receive a clean
+			// error. The connection is still good so we break before the close.
+			dnsErrBufOccured := false
+			var perr *dns.Error
+
+			if errors.As(err, &perr) {
+				if errors.Is(err, dns.ErrBuf) {
+					dnsErrBufOccured = true
+				}
+			}
+
+			if proto == "udp" && ((strings.Contains(err.Error(), "overflow")) || dnsErrBufOccured) {
+				newRet := state.Req.Copy()
+
+				newRet.RecursionAvailable = ret.RecursionAvailable
```

Comment:
IIUC - When a DNS server receives a query, it can respond with the RecursionAvailable bit set to indicate that it supports recursive resolution. So preserving RecursionAvailable field from the original response (ret).

Reply:
Yeah,  But isn't it already copied via `state.Req.Copy()`?

So setting it explicitly here is not necessary?

Reply:
Ah, nevermind - I just realized this is copied from the _request_ - not the _response_...  which is counter intuitive.